Implement strict layering for the images

samueldr opened the pull request:
This stops using the automatic layered approach, instead relying on declaring layers explicitly.

Layers will only include the 𝚫 delta compared to all previous layers.

> **Joshua** Cool. One thing I noticed is that having tons of layers also affects perf, because containerd (docker-like daemon) has to check all of those layers on every deploy

This should help as there is as many layers as declared.

Though this means that care must be taken to optimize sharing between layers. The ordering will matter. This matters if layers are "shared" between customers.

```
┃ ● Layers ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ │
Cmp   Size  Command                                        ├
     34 MB  FROM 1fb5c5a6c7a97cf                           │
     31 MB                                                 │
     52 MB                                                 │
    1.6 MB                                                 │
                                                           │
│ Image Details ├───────────────────────────────────────── │
                                                           │
Image name: layer:kgzpyry823fyqcb9gb79fndx0kg9mz6r         │
Total Image size: 119 MB                                   │
Potential wasted space: 0 B                                │
Image efficiency score: 100 %                              │
                                                           │
```

#### Drawbacks

 - Layers are less likely to be shared between unrelated apps
 - One image per strict layer in the store path

#### Benefits

 - Nix won't rebuild the layers that haven't changed since there's one store path per strict layer.

* * *

Implements #7, solves #9.

jsierles commented:
Nice!

So one thought I had was that if we _can_ do some automated layering, it would be interesting to split up the layers that are more likely to change, and are also the largest. In Rails apps, that will almost always be the gems (one app I have is up to 600MB of gems). So we might benefit from having them split up into smaller layers, if that makes sense.

Also, maybe related, eventually we should make sure we're addressing #8 and not bringing in dev/test gems for comparing layer sizes with other builds.

jsierles commented:
I tested this out - it seems to run quite a bit faster and also deploys faster. But, I'm curious about one detail. It seems like every layer is generated every time the build is run.

In the context of a Dockerfile-based build, lines that haven't changed (representing layers) are skipped completely. So a typical efficient build will only run the build final build step, and only that layer will be pushed.

Is there a way we can mimic this behavior so layers that have not changed from the previous build won't be built or pushed again each time?

Commit:
default.nix: Export ref to ruby and gems from app

Commit:
image-tools: Init with buildSpecifiedLayers

This isn't as "neat" as the automatic layering, as it will produce "fat"
layers, but this allows absolute control over layers.

It will produce one output per layer, but this is both a drawback and a
benefit. At the cost of some storage, we have a "memoized" set of
previous layers the build can start back from.

This means that in the scenario with layers `[ bash ruby app ]`, changing
only the app layer will use the previously built image instead of
rebuilding the whole image again.

The only drawback, compared to `dockerTools.buildLayeredImage` is that
there is no discrete layers re-used across different unrelated builds.
The builds have to be somehow related to have sharing. Though note that
in `[ bash ruby ]`, as long as they are uncustomized, those layers would
be re-used across unrelated images.

Commit:
default.nix: Build using specified layering

samueldr commented:
Just noting that we discussed about the repeated builds concerns, and it might have been a mistake in understanding the output.

Though note that [changing the source dir will affect the gems build](https://github.com/fly-apps/rails-nix/issues/17).
